pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -28,43 +29,144 @@ import (
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
+
 	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
 	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
-// RegisterCallbackFn is the type of the callback function that handlers will provide
-type RegisterCallbackFn func(pluginName string, endpoint string, versions []string, socketPath string) (chan bool, error)
-
 // Watcher is the plugin watcher
 type Watcher struct {
 	path      string
-	handlers  map[string]RegisterCallbackFn
 	stopCh    chan interface{}
 	fs        utilfs.Filesystem
 	fsWatcher *fsnotify.Watcher
 	wg        sync.WaitGroup
-	mutex     sync.Mutex
+
+	mutex       sync.Mutex
+	handlers    map[string]PluginHandler
+	plugins     map[string]pathInfo
+	pluginsPool map[string]map[string]*sync.Mutex // map[pluginType][pluginName]
```

Comment:
there are two type of mutex here:
```
	mutex       sync.Mutex
	pluginsPool map[string]map[string]*sync.Mutex // map[pluginType][pluginName]
```
the mutex is protecting the handler map.
the other is protecting re-registration/de-registration, 
i think both can be renamed that better reflecting the actual meaning.
my suggestion for those two names is: `handlerLock and registrationLocks`, but there could be other option.

Reply:
Would a followup PR be a solution that works for you (the code freeze is tomorrow)?

Reply:
it is fine if you can follow up in a later PR, thanks